modules/internal/ChapelArray.chpl

```diff
@@ -1733,15 +1736,56 @@ module ChapelArray {
        instance of ``val`` in the array, or if ``val`` is not found, a
        tuple containing ``false`` and an unspecified value is returned.
      */
-     @unstable "'Array.find' is unstable"
-     proc find(val: this.eltType): (bool, index(this.domain)) {
+     deprecated "The tuple-returning version of `array.find()` is deprecated; to opt into the new index-returning version, recompile with `-suseNewArrayFind`"
+     proc find(val: this.eltType): (bool, index(this.domain)) where !useNewArrayFind {
       for i in this.domain {
         if this[i] == val then return (true, i);
       }
       var arbInd: index(this.domain);
       return (false, arbInd);
     }
 
+    /*
+
+      Search the array for ``val``, returning the index where it is
+      stored.  If the array contains multiple copies of ``val``, the
+      lexicographically earliest index will be returned.  If ``val``
+      is not found, ``domain.lowBound-1`` is returned instead.  If
+      ``domain.lowBound`` is ``min(idxType)``, the behavior is
+      undefined.
+
+      Support is currently limited to rectangular arrays.
+    */
+    proc find(val: this.eltType) /* TODO: return type is hard to express */ {
```

Comment:
@e-kayrakli : With respect to your comment yesterday:

> [when wanting the most productive type for a given array/domain] we ask the user to write the conditional.

This is where I felt that having a way to say `this.mostConvenientIdxType` (or redefining `this.idxType` to do the same thing) would be more productive than writing the conditional and having it show up in the docs.  Basically, my hope would be that we'd have a compact and standard way to write that type rather than having each such routine have a conditional or rely on its own personal helper function (which was the first way I tackled this before realizing that it made the documentation wonky / require too much indirection).

Reply:
When thinking about how to respond to the design issue for array.find, I found myself needing the same thing without looking at this first :)

I am still not quite happy about providing that new thing because (1) I don't think we can find a name that's less then 3 words, which is too long for the array/domain interface, (2) array interface will have 3 fields (or field-like things) where you can always use any 2 to compute the other 3rd. Which is OK, I guess, but a bit of an overkill and somewhat tautological.

Not that I was vehemently against it, but I'm OK with adding something new, despite my reservations above.

Reply:
@e-kayrakli : I'm not sure I'm interpreting your comment correctly.  Are you saying "OK, I agree we should have this even though naming it will be hard" or "I still don't think we should have this, but I can see why you'd want it?"  If the latter, practically speaking, how would you document this routine's return type in the array library documentation today?

Reply:
Definitely the former. Sorry for not being clear.

Reply:
My best ideas for today:
* `accessType`?  `fullIdxType`?
* Or make the current index type `dimIdxType` and the full thing be `idxType`?

Reply:
Note that Andy has started the discussion on this here: https://github.com/chapel-lang/chapel/issues/19141#issuecomment-1406818593 (Just to avoid getting your ideas lost here)

Reply:
I ended up implementing `fullIdxType` as part of this PR.